### DISCUSSION TITLE
list of mip packages

### DISCUSSION THREAD

**Discussion by mjaspers2mtu** [micropython/micropython#12185] [Libraries & Drivers] [answered]
Hello, after reading through https://docs.micropython.org/en/latest/reference/packages.html
I was left with a question, is there a list of packages that can be installed through mip?
The reason is, I successfully installed ssd1306 through mip, but I have no clue where it got the files from (and I want to view the sourcecode as it only downloaded the .mpy file).
The only clue I have is: https://micropython.org/pi/v2/ but this gives me 403 forbidden when accessed through a browser..

Thanks in advance

**Comment by andrewleech** [comment, marked as answer]
The mip install tool default index is generated automatically from https://github.com/micropython/micropython-lib

The source from all mip-installed packages can be found there, though they're organised in folders like `stdlib` or `micropython` which are abstracted out when mip installing so you might have to check through a couple of directories to find the package you want!

While the docs mention that packages can be installed from other GitHub repositories, the user needs to explicitly specify them if they want to install from anywhere other than the default index.

**Reply by mjaspers2mtu** [reply]
Thank you!

micropython-lib/micropython/drivers/display/ssd1306/ssd1306.py

This was what I was looking for :)

**Comment by bixb922** [comment]
Is the list in https://micropython.org/pi/v2/index.json a complete list of what's in micropython-lib?

**Reply by mjaspers2mtu** [reply]
Oh great you found this! Thanks this is also a nice overview of what can be downloaded through mip :)

**Reply by bixb922** [reply]
Although I'd really like if a little generated table of name/description was linked to https://docs.micropython.org/en/latest/library/index.html...? As simple as a generated one? 

Only after about a year of developing I found that gems like unittest, a more complete json or datetime were present, it did not occur to me that the list in https://docs.micropython.org/en/latest/library/index.html could not be complete... I did not cross check the code back to the documentation... And so I lost a lot of time developing some own code instead of using micropython-lib....

**Comment by jimmo** [comment]
I will add a link and a note about the index.json in the packages.rst docs. The docs do already reference "micropython-lib" as the source of packages, but it could probably be more explicit (and the glossary entry could be improved).

With regard to publishing packages, adding all micropython-lib packages to the docs is something that we'd like to do. I'm right in the middle of adding support for auto-generated docs to our docs build (and migrating existing API docs to Python stub files). Adding micropython-lib packages will not be too difficult after this.

**Reply by bixb922** [reply]
Excellent, thanks for all the good work!

**Comment by CombatVolcano** [comment]
I need to install this: https://github.com/Thomascountz/micropython_i2c_lcd
I can't for the life of me figure out how to get MIP to point at that
What NAME do I use for the package? What URL? It doesn't accept anything
It is connected to wifi

**Comment by andrewleech** [comment]
@CombatVolcano it just so happens at the micropython meetup this week a couple of community memebers started putting together an indexer for mip packages! it's early days but you can take a look at a preview here: https://mim.oliverrobson.tech/

**Reply by CombatVolcano** [reply]
Cool. I don't know why that was necessary and never existed before, but what I need to know is how to get MIP to install from github.

What syntax does this thing expect? And this package I want, it says in the repo that it was uploaded to MIP but how the heck am I supposed to find that name?

I have to admit I am about to drop a day's work getting all this running and try circuitpython instead because this is a MESS!

Seriously the most important thing about python is being able to install packages and all of the documentation is USELESS when it comes to functionally doing that.

I highly suggest you pretend you know very little and go read the documentation and try to figure out how to get MIP to install anything. You won't be able to do it. I mean, please, put me to shame. Link me to the page I can't find that explains it all with actual examples and not /psuedo/code/this/is/an/example in everything.

What path format does MIP expect because nothing I try works!

**Reply by andrewleech** [reply]
Oh sorry, at first glance I thought you were asking about the top post of this discussion, seeing a list of everything available. 

The docs mention installing py files directly from GitHub, https://docs.micropython.org/en/latest/reference/packages.html in the section referring to `github:`

eg. 
```
>>> mip.install("github:org/repo/path/foo.py")
```

So to install a file from that repo it should be something like:
```
>>> import mip
>>> mip.install("github:Thomascountz/micropython_i2c_lcd/hd44780.py")
Downloading github:Thomascountz/micropython_i2c_lcd/hd44780.py to /home/corona/.micropython/lib
Copying: /home/corona/.micropython/lib/hd44780.py
Done
>>>
```

It's possible for a repo to include a mip specific json file you can point mip at to install everything in one go, but that's pretty new so not many libraries are using it yet.

**Reply by CombatVolcano** [reply]
Thank you very much for your help. The format you provided does indeed work. I do blame myself, for not being able to translate what was in the documentation into a functional URL, but it would really, really help if there were a clearer explanation there as to how this is intended to function.

The list of packages is certainly a nice thing to have... though the metadata really does need improving.

for anyone else that finds this in frustration, I posted code to get you to hello world on an LCD package [here](https://github.com/Thomascountz/micropython_i2c_lcd/issues/6) and it shows how to MIP install a package from github (which may not be properly configured on github)

**Reply by andrewleech** [reply]
> The list of packages is certainly a nice thing to have... though the metadata really does need improving.

Yes the metadata been the discussion so far, this is all work in progress and people are figuring out how to get more data to show

**Reply by Josverl** [reply]
A I have been working on generating documentation pages based on stubs, 

As part of that work I also did a quick test to see if the same approach can be used to generate docpages for modules in `micropython-lib`
That was not too difficult , there are a pile of .rST formatting errors to correct, that actually took most of the time.

I added a note how to install, that can/should also be linked though to the mpremote docs.

https://josverl.github.io/autodoc201/modules/base64/index.html
![image](https://github.com/user-attachments/assets/72657dee-ff3f-4bf0-9090-b549fa565252)

I have setup 3 simple index pages that really should have a simple description as well,
so would love to collaborate on that with @oliverrobson 

I tried to re-use the same structure that is in micropython-lib but ran into some naming collisions with the exiting documentation.

![image](https://github.com/user-attachments/assets/709f9733-7d9e-418f-8563-eefe6a5859c0)

- [micropython](https://josverl.github.io/autodoc201/mpy-lib/micropython.html)
- [python-stdlib](https://josverl.github.io/autodoc201/mpy-lib/python-stdlib.html)
- [python-ecosys](https://josverl.github.io/autodoc201/mpy-lib/python-community.html)

Note: 
    Some modules are missing for the above due to the simple naive module parser/copy in the current PoC) 
    Currently the install hint requires a module docstring of at least one character.

Repo: https://github.com/Josverl/autodoc201

**Comment by peterhinch** [comment]
@CombatVolcano The above discussion is about official libraries while the [one you want to install](https://github.com/Thomascountz/micropython_i2c_lcd) is a user-contributed library.

Unfortunately that library is not mip-installable because it lacks a `package.json` file which tells MIP what files to download. You'll need to clone the repo to your PC (`git clone https://github.com/Thomascountz/micropython_i2c_lcd`) then copy the files to your target. It would be good if authors of libraries could be persuaded to create this file, but we're relying on knowledge and goodwill here :)

I wonder if the work to create an index will include user-contributed MIP installable libraries?

**Reply by Josverl** [reply]
Hi Peter,
I agree with you that there should be a simple way to include mip installable modules from the ecosystem.
As I think a lot (but not all) of these are already on awesome-micropython, and there was a thought to add a mip installable badge there,
that would be a good combination.

Im pretty sure that given
-  a list of URLs 
- a simple rule set bases on  current MIP requirements and common practices for authorship and licencing
- Optionally:  a simple description.md/.rst or a typestub.

That I can get Sphinx to render some decent basic documentation across all these modules,
including cross references to the micropython and python documentation
